### PR TITLE
fix: Fixed instrumenting the same module installed in different locations

### LINKED ISSUE #2033: Restore ability to instrument the same module but in different folder locations
## Is your feature request related to a problem? Please describe.
In #1563 we added the ability to instrument the same module more than once.  However there were some issues with that. We shifted where we flagged if a module was already instrumented from the export to the instrumentation hooks.  This surfaced a different issue where you could not instrument multiple version of the same module(#1799).  However, that still had an issue where you could have the same module and version installed in different locations.  The modern npm resolution typically flattens directories to where this is not possible but you could have a module installed in different roots. One could argue this is an anti-pattern but it still works.  We need to [restore this ability](https://github.com/newrelic/node-newrelic/pull/1563/files#diff-3d0c4388e915f657046d7040af798b9376a45f953db048c457fe98a148d22b10L559), which essentially should be storing the instrumentation flag back on the exported module. But instead of having a boolean flag we need to keep track of which instrumentation hooks were registered.

## Feature Description
Support instrumenting the same module and version in different folder locations.

### PR DESCRIPTION
**Pull request by jsumners-nr**
This PR resolves #2033.

Before we can merge this PR, we need to:

- [x] Merge and release https://github.com/newrelic/node-test-utilities/pull/195
- [x] Update https://github.com/newrelic/node-newrelic-aws-sdk/pull/279 and release after step 1 is done
- [x] Update references to those modules in this branch

### CODE REVIEW

File: lib/shimmer.js [hunk 1]
```diff
@@ -555,18 +572,13 @@ function instrumentPostLoad(agent, nodule, moduleName, resolvedName, esmResolver
...
-      shim,
-      instrumentation.specifier || moduleName,
-      NAMES.FEATURES.INSTRUMENTATION.ON_REQUIRE
-    )
+    trackInstrumentationUsage(agent, shim, moduleName, NAMES.FEATURES.INSTRUMENTATION.ON_REQUIRE)
 
     // Tracking instrumentation is only used to add the supportability metrics
     // that occur directly above this.  No reason to attempt to load instrumentation
     // as it does not exist.
     if (instrumentation.type === MODULE_TYPE.TRACKING) {
```
Comment (bizob2828) [inline lib/shimmer.js:580]: is it a problem we have these types defined in two places now? shim constants and instrumentation descriptor?

File: test/versioned-external/external-repos.js [hunk 2]
```diff
@@ -14,8 +14,8 @@
 const repos = [
   {
     name: 'aws-sdk',
-    repository: 'https://github.com/newrelic/node-newrelic-aws-sdk.git',
-    branch: 'main'
+    repository: 'https://github.com/jsumners-nr/node-newrelic-aws-sdk.git',
```
Comment (bizob2828) [inline test/versioned-external/external-repos.js:17]: this can be restored now correct?

**Review by bizob2828 — CHANGES_REQUESTED** [review]
A few comments but hard to see the state of this atm because there's a merge conflict. I'll continue review after that is addressed

File: test/integration/agent/reinstrument.tap.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      lines.push(chunk.toString())
+      cb()
+    }
+  })
+  const person = new Person(dest)
+  console.log(person.isHuman)
+  console.log(lines)
+
+  // This should be true if we have solved the problem correctly.
+  console.log(didWeDoIt.length === 2)
```
[the remark this answers is not in the document]
Comment (jsumners-nr) [reply test/integration/agent/reinstrument.tap.js:40]: Basically, I forgot to codify this into an actual test. A consequence of the duration.

File: lib/shimmer.js (hunk 1 above)
Comment (jsumners-nr) [reply lib/shimmer.js:580]: I meant to clean this up.

Comment (jsumners-nr) [reply lib/shimmer.js]: I have addressed this in [`ca26505` (#2104)](https://github.com/newrelic/node-newrelic/pull/2104/commits/ca26505864a07bfb378a07d8992e3d84434313c4)

File: lib/shimmer.js [hunk 4]
```diff
@@ -534,16 +550,17 @@ function applyDebugState(shim, nodule, inEsm) {
 function instrumentPostLoad(agent, nodule, moduleName, resolvedName, esmResolver) {
   // default to Node.js version, this occurs for core libraries
   const pkgVersion = resolvedName ? shimmer.getPackageVersion(moduleName) : process.version
```
[the remark this answers is not in the document]
Comment (jsumners-nr) [reply lib/shimmer.js]: Good catch.

File: lib/shimmer.js [hunk 5]
```diff
@@ -467,23 +473,26 @@ const shimmer = (module.exports = {
...
-    )
-    return didInstrument.length === instrumentations.length
+  isInstrumented(moduleName, resolvedName) {
+    const items = shimmer.registeredInstrumentations
+      .getAllByName(moduleName)
+      .filter(
+        (item) =>
+          item.instrumentation.resolvedName === resolvedName && item.meta.instrumented === true
+      )
+    return items.length > 0
```
[the remark this answers is not in the document]
Comment (jsumners-nr) [reply lib/shimmer.js]: I must've gotten lost around the purpose of this function. I had to update it a few times.

**Review by bizob2828 — APPROVED** [review]
Great work. I know this was one was a tough one. Great job abstacting away the tracking and registration of instrumentation.  I verified the new test proves the intent of this PR. I also ran the versioned tests with security agent.

**Pull request merged**